beacon/goclient/goclient.go

```diff
@@ -190,43 +234,116 @@ func New(
+		eth2clienthttp.WithTimeout(commonTimeout),
+		eth2clienthttp.WithReducedMemoryUsage(true),
+		eth2clienthttp.WithAllowDelayedStart(true),
+	)
 	if err != nil {
-		logger.Error(clResponseErrMsg,
-			zap.String("api", "NodeVersion"),
+		logger.Error("Consensus http client initialization failed",
+			zap.String("address", addr),
 			zap.Error(err),
 		)
-		return nil, fmt.Errorf("failed to get node version: %w", err)
+
+		return nil, fmt.Errorf("create http client: %w", err)
 	}
-	if nodeVersionResp == nil {
-		logger.Error(clNilResponseErrMsg,
-			zap.String("api", "NodeVersion"),
-		)
-		return nil, fmt.Errorf("node version response is nil")
+
+	return httpClient.(*eth2clienthttp.Service), nil
+}
+
+// assertSameSpec should receive a non-empty list
+func assertSameSpec(ctx context.Context, services ...Client) error {
+	firstSpec, err := services[0].Spec(ctx, &api.SpecOpts{})
+	if err != nil {
+		return fmt.Errorf("get first spec: %w", err)
 	}
-	client.nodeVersion = nodeVersionResp.Data
-	client.nodeClient = ParseNodeClient(nodeVersionResp.Data)
-
-	logger.Info("consensus client connected",
-		fields.Name(httpClient.Name()),
-		fields.Address(httpClient.Address()),
-		zap.String("client", string(client.nodeClient)),
-		zap.String("version", client.nodeVersion),
-	)
 
-	go client.registrationSubmitter(slotTickerProvider)
-	// Start automatic expired item deletion for attestationDataCache.
-	go client.attestationDataCache.Start()
+	firstGenesis, err := services[0].Genesis(ctx, &api.GenesisOpts{})
+	if err != nil {
+		return fmt.Errorf("get first genesis: %w", err)
+	}
 
-	return client, nil
+	for _, service := range services[1:] {
+		srvSpec, err := service.Spec(ctx, &api.SpecOpts{})
+		if err != nil {
+			return fmt.Errorf("get service spec: %w", err)
+		}
+
+		srvGenesis, err := service.Genesis(ctx, &api.GenesisOpts{})
+		if err != nil {
+			return fmt.Errorf("get service genesis: %w", err)
+		}
+
+		if err := sameSpec(firstSpec.Data, srvSpec.Data); err != nil {
+			return fmt.Errorf("different spec: %w", err)
+		}
+
+		if err := sameGenesis(firstGenesis.Data, srvGenesis.Data); err != nil {
+			return fmt.Errorf("different genesis: %w", err)
+		}
+	}
+
+	return nil
 }
 
-func (gc *GoClient) nodeSyncing(ctx context.Context, opts *api.NodeSyncingOpts) (*api.Response[*apiv1.SyncState], error) {
-	return gc.client.NodeSyncing(ctx, opts)
+func sameSpec(a, b map[string]any) error {
+	paramsToCheck := []string{
+		"CONFIG_NAME",
+		"CAPELLA_FORK_VERSION",
+		"SECONDS_PER_SLOT",
+		"SLOTS_PER_EPOCH",
+		"EPOCHS_PER_SYNC_COMMITTEE_PERIOD",
+		"SYNC_COMMITTEE_SIZE",
+		"SYNC_COMMITTEE_SUBNET_COUNT",
+		"TARGET_AGGREGATORS_PER_COMMITTEE",
+		"TARGET_AGGREGATORS_PER_SYNC_SUBCOMMITTEE",
+		// NOTE: not checking "INTERVALS_PER_SLOT" because it's not set on some clients
+	}
+
+	for _, param := range paramsToCheck {
+		if a[param] != b[param] {
+			return fmt.Errorf("param %s mismatch, got %v and %v", param, a[param], b[param])
+		}
+	}
+
+	return nil
 }
```

Comment:
i dont know the benefit of checking all of these

checking just `GENESIS_FORK_VERSION` should be sufficient, as it's supposed to be the "network ID"

Reply:
i see that this variable is in genesis data, so essentially sameSpec isnt necessary

Reply:
I agree, done